lib/PuppeteerSharp/Helpers/TempDirectory.cs

```diff
@@ -29,37 +29,44 @@ private TempDirectory(string path)
             Path = path;
         }
 
-        ~TempDirectory()
-        {
-            DisposeCore();
-        }
+        ~TempDirectory() => Dispose();
 
         public string Path { get; }
 
-        public void Dispose()
+#pragma warning disable CA1816
+        public void Dispose() => _ = DisposeAsync();
+#pragma warning restore CA1816
+
+        public override string ToString() => Path;
+
+        /// <inheritdoc />
+        public async ValueTask DisposeAsync()
         {
+            lock (_disposingLock)
+            {
+                _deleteTask ??= DeleteAsync();
+            }
```

Comment:
TLDR; using a lock  and ensuring we have a single task of `DeleteAsync` looks good.

My thoughts while looking at this:

When looking at other implementations of `DisposeAsync` we have different mindsets about how to handle re-entrance.

`TaskQueue.DisposeAsync` returns immediately on re-entrance and in a thread-safe manner.
https://github.com/hardkoded/puppeteer-sharp/blob/7d7e6f01e5af831d1b03986746d00af04b311da8/lib/PuppeteerSharp/Helpers/TaskQueue.cs#L32-L35

`DisposableTasksStack.DisposeAsync` also immediately returns on re-entrance, but in thread-unsafe manner, which seems fine when looking at its two usages.
https://github.com/hardkoded/puppeteer-sharp/blob/7d7e6f01e5af831d1b03986746d00af04b311da8/lib/PuppeteerSharp/DisposableTasksStack.cs#L44-L58

`ExecutionContext.DisposeAsync` on the other hand does not have special re-entrance handling.
https://github.com/hardkoded/puppeteer-sharp/blob/7d7e6f01e5af831d1b03986746d00af04b311da8/lib/PuppeteerSharp/ExecutionContext.cs#L72-L90

We for sure don't want multiple _different_ invocations of `DeleteAsync`.
I added the thread-safe variant using `Interlocked` in 45ed534343989ee358bd6f8d90bdf564924aab9b.
I'm not sure we _really_ need the thread-safety, but I suggest we keep it thread-safe.

Since `DeleteAsync` does sync IO (`Directory.Exists` and `Directory.Delete`) before reaching its first `await`, calling it inside a `lock` could be sub-optimal if the lock is expected to be congested.

So the remaining question I see is: how should `TempDirectory.DisposeAsync` behave on re-entrance?

Can re-entrant callers return immediately, expecting the first caller to complete the cleanup?
```cs
private int _disposed;
...

public async ValueTask DisposeAsync()
{
    if (Interlocked.Exchange(ref _disposed, 1) != 0)
    {
        return;
    }

    await DeleteAsync().ConfigureAwait(false);
    _tcs.SetResult(true);
    GC.SuppressFinalize(this);
}
```

Do concurrent callers need to still await the call to the single `DeleteAsync` task to ensure that `Path` has been deleted when returning?

```cs
private readonly TaskCompletionSource<bool> _tcs = new();
private int _disposed;
...

public async ValueTask DisposeAsync()
{
    if (Interlocked.Exchange(ref _disposed, 1) != 0)
    {
        await _tcs.Task.ConfigureAwait(false);
        return;
    }

    await DeleteAsync().ConfigureAwait(false);
    _tcs.SetResult(true);
    GC.SuppressFinalize(this);
}
```

Reply:
> TLDR; using a lock and ensuring we have a single task of `DeleteAsync` looks good.
> 
> My thoughts while looking at this:
> 
> When looking at other implementations of `DisposeAsync` we have different mindsets about how to handle re-entrance.

I think that, functionally speaking, all the calls to `DisposeAsync` should return the same task. But I couldn't find any code in `dotnet/runtime` to back me up.
That left me thinking that we are breaking some pattern here. 

The good news is that we could just simply call `DeleteAsync` instead of `DisposeAsync`.
So, I'm removing `IDisposableAsync`. The Launcher should call `DeleteAsync`. If everything goes as expected, the directory should be gone when disposed.

Thanks for your help man.